src/crimson/os/seastore/journal/segmented_journal.cc

```diff
@@ -93,48 +94,115 @@ SegmentedJournal::prep_replay_segments(
     }
   });
 
-  auto journal_tail = segments.rbegin()->second.journal_tail;
-  segment_provider.update_journal_tail_committed(journal_tail);
-  auto replay_from = journal_tail.offset;
-  auto from = segments.begin();
-  if (replay_from != P_ADDR_NULL) {
-    from = std::find_if(
+  auto last_segment_id = segments.rbegin()->first;
+  auto last_header = segments.rbegin()->second;
+  segment_provider.update_journal_tails(
+      last_header.dirty_tail, last_header.alloc_tail);
+  auto seq = journal_seq_t{
+    last_header.segment_seq,
+    paddr_t::make_seg_paddr(last_segment_id, 0)
+  };
+  INFO("scanning {} for journal tail deltas", seq);
+  return seastar::do_with(
+    scan_valid_records_cursor(seq),
+    SegmentManagerGroup::found_record_handler_t(
+      [FNAME, this](
+        record_locator_t locator,
+        const record_group_header_t& record_group_header,
+        const bufferlist& mdbuf
+      ) -> SegmentManagerGroup::scan_valid_records_ertr::future<>
+    {
+      DEBUG("decoding {} at {}", record_group_header, locator);
+      bool has_tail_delta = false;
+      auto maybe_headers = try_decode_record_headers(
+          record_group_header, mdbuf);
+      if (!maybe_headers) {
+        // This should be impossible, we did check the crc on the mdbuf
+        ERROR("unable to decode headers from {} at {}",
+              record_group_header, locator);
+        ceph_abort();
+      }
+      for (auto &record_header : *maybe_headers) {
+        ceph_assert(is_valid_transaction(record_header.type));
+        if (is_cleaner_transaction(record_header.type)) {
+          has_tail_delta = true;
+        }
+      }
+      if (has_tail_delta) {
+        bool found_delta = false;
+        auto maybe_record_deltas_list = try_decode_deltas(
+          record_group_header, mdbuf, locator.record_block_base);
+        if (!maybe_record_deltas_list) {
+          ERROR("unable to decode deltas from {} at {}",
+                record_group_header, locator);
+          ceph_abort();
+        }
+        for (auto &record_deltas : *maybe_record_deltas_list) {
+          for (auto &[ctime, delta] : record_deltas.deltas) {
+            if (delta.type == extent_types_t::JOURNAL_TAIL) {
+              found_delta = true;
+              journal_tail_delta_t tail_delta;
+              decode(tail_delta, delta.bl);
+              if (tail_delta.alloc_tail == JOURNAL_SEQ_NULL) {
+                tail_delta.alloc_tail = locator.write_result.start_seq;
+              }
+              if (tail_delta.dirty_tail == JOURNAL_SEQ_NULL) {
+                tail_delta.dirty_tail = locator.write_result.start_seq;
+              }
+              segment_provider.update_journal_tails(
+                  tail_delta.dirty_tail, tail_delta.alloc_tail);
+            }
+          }
+        }
+        ceph_assert(found_delta);
+      }
+      return seastar::now();
+    }),
+    [this, nonce=last_header.segment_nonce](auto &cursor, auto &handler)
+  {
```

Comment:
Formatting here is inconsistent, { should probably be on the previous line.  Alternately, it might be worth splitting out a helper.

Reply:
> Formatting here is inconsistent, { should probably be on the previous line.

I'm trying not to introduce too many indentions to the `do_with()` body function block, which is 2 spaces (instead of 4) to the outer `return seastar::do_with(`. So it's easier to understand the nest level by counting spaces or the visual aligment.

The `{` here with a newline is to match the indention of the below `}`, I thought it can improve readability by matching the start and the end of the code block with  nesting logics. Otherwise, there is no visual stopper until `return seastar::do_with(` above.

How do you think? @athanatos 

> Alternately, it might be worth splitting out a helper.

Yeah, I think so.

Reply:
just like:
https://github.com/ceph/ceph/blob/dcd0541bb2cdf56452d4450dc1a8642b023d5e64/src/crimson/os/seastore/segment_manager_group.cc#L330-L349

instead of:
https://github.com/ceph/ceph/blob/main/src/crimson/os/seastore/transaction_manager.cc#L112-L135

Reply:
> Alternately, it might be worth splitting out a helper.

done.

Reply:
lgtm